### PR TITLE
lr_schedulers: add get_polynomial_decay_schedule_with_warmup

### LINKED ISSUE #6070: lightning_base: new clarg: lr_scheduler=polynomial_decay
lr_scheduler should be a string with many options, including polynomial_decay, like this [command](https://github.com/pytorch/fairseq/blob/master/examples/mbart/README.md#finetune-on-en-ro)

### PR DESCRIPTION
**Pull request by stas00**
this PR adds a new scheduler plus test, code is based on amalgamation of a few different implementations

I'm not sure it's 100% correct - needs more experimenting - but feedback is welcome.

For reference here are 3 different implementations of this scheduler:
1. https://github.com/pyprob/pyprob/blob/master/pyprob/nn/inference_network.py#L357
2. https://github.com/cmpark0126/pytorch-polynomial-lr-decay/blob/master/torch_poly_lr_decay/torch_poly_lr_decay.py#L5
3. https://github.com/pytorch/fairseq/blob/master/fairseq/optim/lr_scheduler/polynomial_decay_schedule.py - this one has an extra feature `--force-anneal`

@sshleifer

### CODE REVIEW

File: src/transformers/optimization.py [hunk 1]
```diff
@@ -165,6 +165,52 @@ def lr_lambda(current_step):
...
+            Power factor.
+        last_epoch (:obj:`int`, `optional`, defaults to -1):
+            The index of the last epoch when resuming training.
+
+    Return:
+        :obj:`torch.optim.lr_scheduler.LambdaLR` with the appropriate schedule.
+
+    """
+
+    lr_init = optimizer.defaults["lr"]
```
Comment (sshleifer) [inline src/transformers/optimization.py:196]: assert lr_init > lr_end?
I think we are starting at 3e-5 so 1e-5 feels a little high, but we should just copy fairseq.

**Review by sshleifer — APPROVED** [review]
LGTM, pending comment.

File: src/transformers/optimization.py (hunk 1 above)
Comment (stas00) [reply src/transformers/optimization.py:196]: theirs is 0.0:
``` 
parser.add_argument('--end-learning-rate', default=0.0, type=float)
```

Comment (stas00) [reply src/transformers/optimization.py:196]: I'm also not sure about `lr_init = optimizer.defaults["lr"]` - what if there are groups of `lrs` that aren't the same. The problem with `LambdaLR` - the `base_lr` of the layer isn't passed to `LambdaLR` and therefore this class is not set to deal with different `lrs`.

File: src/transformers/optimization.py [hunk 2]
```diff
@@ -165,6 +165,52 @@ def lr_lambda(current_step):
...
+
+    def lr_lambda(current_step: int):
+        if current_step < num_warmup_steps:
+            return float(current_step) / float(max(1, num_warmup_steps))
+        elif current_step > num_training_steps:
+            return lr_end / lr_init  # as LambdaLR multiplies by lr_init
+        else:
+            lr_range = lr_init - lr_end
+            decay_steps = num_training_steps - num_warmup_steps
+            num_updates = current_step + 1
```
Comment (stas00) [inline src/transformers/optimization.py:206]: I'm also not sure about this one. I was trying to fix the one-off case to match the output of this implementation of the same: https://github.com/cmpark0126/pytorch-polynomial-lr-decay/blob/master/torch_poly_lr_decay/torch_poly_lr_decay.py#L5 - which I was using to verify my implementation - my schedule was off by one place in the sequence, so this +1 made it match. But it could be that the other one isn't perfect.

In the big picture of hundreds of steps it probably doesn't matter match, but if can make it precise, why not.

File: src/transformers/optimization.py (hunk 1 above)
Comment (stas00) [reply src/transformers/optimization.py:196]: Assert added.

File: src/transformers/optimization.py [hunk 3]
```diff
@@ -165,6 +165,53 @@ def lr_lambda(current_step):
...
+    initial lr set in the optimizer.
+
+    Args:
+        optimizer (:class:`~torch.optim.Optimizer`):
+            The optimizer for which to schedule the learning rate.
+        num_warmup_steps (:obj:`int`):
+            The number of steps for the warmup phase.
+        num_training_steps (:obj:`int`):
+            The total number of training steps.
+        lr_end (:obj:`float`, `optional`, defaults to 1e-08):
```
Comment (sgugger) [inline src/transformers/optimization.py:184]:
```suggestion
        lr_end (:obj:`float`, `optional`, defaults to 1e-7):
```
(Does not match the value in the definition.)

File: src/transformers/optimization.py [hunk 4]
```diff
@@ -165,6 +165,53 @@ def lr_lambda(current_step):
...
+            The end LR.
+        power (:obj:`float`, `optional`, defaults to 1.0):
```
Comment (sgugger) [inline src/transformers/optimization.py:186]:
```suggestion
        power (:obj:`float`, `optional`, defaults to 2.0):
```
Same.

File: src/transformers/optimization.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/optimization.py:207]: This does seem weird. I'm not super experienced with the `LambdaLR` class, but it seems from a quick browse that the number of steps has been already incremented here, so the +1 is not needed.

**Review by sgugger — APPROVED** [review]
Thanks for adding this @stas00 !

**Review by LysandreJik — APPROVED** [review]
Nice! Maybe you can rebase on `master` to get your own PR's test refactor, so that you can implement it here for the additional schedule? Will prevent from doing another PR to refactor this one.

File: src/transformers/optimization.py (hunk 4 above)
Comment (stas00) [reply src/transformers/optimization.py:186]: thank you for catching those inconsistencies, @sgugger .

fairseq implementation's default is actually power=1.0 and not 2.0, whereas others use 2.0. 
https://github.com/pytorch/fairseq/blob/master/fairseq/optim/lr_scheduler/polynomial_decay_schedule.py#L37

power=1.0 kind of defeats the purpose behind polynomial. 

Which do you recommend we use as the default? Especially since we currently don't have the cl arg to override the default if it's used for example scripts.

**Comment by stas00** [comment]
All done, just need to decide whether to use the default of 1.0 for `power` as in fairseq, or 2.0 (or another value) as it actually does something polynomial.

**Comment by stas00** [comment]
I run fairseq as recommended [here](https://github.com/pytorch/fairseq/blob/master/examples/mbart/README.md#finetune-on-en-ro) and no, it is using the default power=1.0

I double checked their code, it doesn't get overriden anywhere:
```
fairseq/optim/lr_scheduler/polynomial_decay_schedule.py:        self.power = args.power
fairseq/optim/lr_scheduler/polynomial_decay_schedule.py:        parser.add_argument('--power', default=1.0, type=float)
fairseq/optim/lr_scheduler/polynomial_decay_schedule.py:        print("POWER:", self.power)
fairseq/optim/lr_scheduler/polynomial_decay_schedule.py:            lr = lr_range * pct_remaining ** (self.power) + self.end_learning_rate
```

I will open an issue there and report back.

File: src/transformers/optimization.py (hunk 4 above)
Comment (stas00) [reply src/transformers/optimization.py:186]: Agreed. and also please see https://github.com/huggingface/transformers/pull/6361#issuecomment-672287450.

may be we can merge at 2.0 and adjust later if need be. this is a default after all.

Comment (sgugger) [reply src/transformers/optimization.py:186]: Agreed.

**Comment by sshleifer** [comment]
👍
